src/interface/swarm_pack.hpp

```diff
@@ -0,0 +1,207 @@
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SWARM_PACK_HPP_
+#define INTERFACE_SWARM_PACK_HPP_
+
+#include <algorithm>
+#include <functional>
+#include <limits>
+#include <map>
+#include <memory>
+#include <regex>
+#include <set>
+#include <string>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "coordinates/coordinates.hpp"
+#include "interface/mesh_data.hpp"
+#include "interface/meshblock_data.hpp"
+#include "interface/swarm.hpp"
+#include "interface/swarm_pack_base.hpp"
+#include "interface/variable.hpp"
+#include "utils/concepts_lite.hpp"
+#include "utils/utils.hpp"
+
+namespace parthenon {
+
+class SwarmPackIdx {
+ public:
+  KOKKOS_INLINE_FUNCTION
+  explicit SwarmPackIdx(std::size_t var_idx) : vidx(var_idx), offset(0) {}
+  KOKKOS_INLINE_FUNCTION
+  SwarmPackIdx(std::size_t var_idx, int off) : vidx(var_idx), offset(off) {}
+
+  KOKKOS_INLINE_FUNCTION
+  SwarmPackIdx &operator=(std::size_t var_idx) {
+    vidx = var_idx;
+    offset = 0;
+    return *this;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  std::size_t VariableIdx() { return vidx; }
+  KOKKOS_INLINE_FUNCTION
+  int Offset() { return offset; }
+
+ private:
+  std::size_t vidx;
+  int offset;
+};
+
+// Operator overloads to make calls like `my_pack(b, my_pack_idx + 3, k, j, i)` work
+template <class T, REQUIRES(std::is_integral<T>::value)>
+KOKKOS_INLINE_FUNCTION SwarmPackIdx operator+(SwarmPackIdx idx, T offset) {
+  return SwarmPackIdx(idx.VariableIdx(), idx.Offset() + offset);
+}
+
+template <class T, REQUIRES(std::is_integral<T>::value)>
+KOKKOS_INLINE_FUNCTION SwarmPackIdx operator+(T offset, SwarmPackIdx idx) {
+  return idx + offset;
+}
+
+// Namespace in which to put variable name types that are used for indexing into
+// SwarmPack<[type list of variable name types]> on device
+namespace swarm_variable_names {
+// Struct that all variable_name types should inherit from
+template <bool REGEX, typename T = Real, int... NCOMP>
+struct base_t {
+  using data_type = T;
+
+  KOKKOS_INLINE_FUNCTION
+  base_t() : idx(0) {}
+
+  KOKKOS_INLINE_FUNCTION
+  explicit base_t(int idx1) : idx(idx1) {}
+
+  virtual ~base_t() = default;
+
+  // All of these are just static methods so that there is no
+  // extra storage in the struct
+  static std::string name() {
+    PARTHENON_FAIL("Need to implement your own name method.");
+    return "error";
+  }
+  KOKKOS_INLINE_FUNCTION
+  static bool regex() { return REGEX; }
+  KOKKOS_INLINE_FUNCTION
+  static int ndim() { return sizeof...(NCOMP); }
+  KOKKOS_INLINE_FUNCTION
+  static int size() { return multiply<NCOMP...>::value; }
+
+  const int idx;
+};
+
+} // namespace swarm_variable_names
+
+template <typename TYPE, class... Ts>
+class SwarmPack : public SwarmPackBase<TYPE> {
```

Comment:
I wonder if we could just extend/template `SparsePack` to provide the functionality given by `SwarmPack` and `SparsePack` (although the would have to inherit different from different base classes). Reduces code duplication but may result in some gross SFINAE.

Reply:
resolving this for now, although I think this would be nice to do in a future PR.